grails-app/domain/org/pih/warehouse/invoice/InvoiceItem.groovy

```diff
@@ -191,6 +191,39 @@ class InvoiceItem implements Serializable {
     Integer getQuantityAvailableToInvoice() {
         return shipmentItem ? (shipmentItem.quantityToInvoiceInStandardUom / quantityPerUom) + quantity : null
     }
+  
+    InvoiceItem clone() {
+        InvoiceItem clone = new InvoiceItem(
+                invoice: invoice,
+                product: product,
+                glAccount: glAccount,
+                budgetCode: budgetCode,
+                quantity: quantity,
+                quantityUom: quantityUom,
+                quantityPerUom: quantityPerUom,
+                amount: amount,
+                unitPrice: unitPrice,
+                inverse: inverse,
+        )
+
+        // A cloned invoice item retains all the same shipment items, and order items and adjustments as the original
+        // invoice. We also make sure to update the other side of the relationship to include the newly cloned invoice
+        // item since the relationship is bidirectional.
+        for (ShipmentItem shipmentItem : shipmentItems) {
+            clone.addToShipmentItems(shipmentItem)
+            shipmentItem.addToInvoiceItems(clone)
+        }
+        for (OrderItem orderItem : orderItems) {
+            clone.addToOrderItems(orderItem)
+            orderItem.addToInvoiceItems(clone)
+        }
+        for (OrderAdjustment orderAdjustment : orderAdjustments) {
+            clone.addToOrderAdjustments(orderAdjustment)
+            orderAdjustment.addToInvoiceItems(clone)
+        }
```

Comment:
Can't it be just `shipmentItem.setInvoicetems()` instead of looping over all items?

+ I am not sure how it works in Grails, but is there a need to copy the relation for both sides? If you set it only for the "owner" of the relation, it will also be set for the second object, but after closing the session? @kchelstowski is a hibernate master, so probably he should know more about that :smile:

Reply:
good point. I'm also very curious about this :eyes:

Reply:
@alannadolny @EWaterman this version is fine, because we loop through the shipmentItems, not the invoiceItems, so for each SHIPMENT item we set the clone.
The same with the loops below.